bootstrap/controllers/machine_deployment_controller.go

```diff
@@ -0,0 +1,437 @@
+	PatchHelper       *patch.Helper
+	UpgradeTo         string
+	OwnedMachines     []*clusterv1.Machine
+}
+
+// NewMachineDeploymentReconciler creates a new MachineDeploymentReconciler.
+func (r *MachineDeploymentReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.scheme = mgr.GetScheme()
+	r.recorder = mgr.GetEventRecorderFor("ck8s-machine-deployment-controller")
+
+	if r.machineGetter == nil {
+		r.machineGetter = &ck8s.Management{
+			Client: r.Client,
+		}
+	}
+
+	// NOTE(Hue): Initially, I tried to go with comprehensive predicates but there was two problems with that:
+	// 1. It was not really understandable and mantainable.
+	// 2. Sometimes the reconciliation was not getting triggered when it should have, debugging this
+	// through the predicates was a nightmare.
+	if err := ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.MachineDeployment{}).
+		Owns(&clusterv1.Machine{}).
+		Complete(r); err != nil {
+		return fmt.Errorf("failed to get new controller builder: %w", err)
+	}
+
+	return nil
+}
+
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters;clusters/status,verbs=get;list;watch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machines;machines/status,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machinesets;machinesets/status,verbs=get;list;watch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machinedeployments;machinedeployments/status,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile handles the reconciliation of a MachineDeployment object.
+func (r *MachineDeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// NOTE(Hue): Generate a tarce ID for this reconciliation to track logs better.
+	traceID := trace.NewID()
+	log := r.Log.WithValues("machine_deployment", req.NamespacedName, "trace_id", traceID)
+	log.Info("Reconciliation started...")
+
+	machineDeployment := &clusterv1.MachineDeployment{}
+	if err := r.Get(ctx, req.NamespacedName, machineDeployment); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("MachineDeployment resource not found. Ignoring since object must be deleted.")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get MachineDeployment: %w", err)
+	}
+
+	if !machineDeployment.DeletionTimestamp.IsZero() {
+		log.Info("MachineDeployment is being deleted, skipping reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	ownedMachines, err := r.getOwnedMachines(ctx, machineDeployment)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get owned machines: %w", err)
+	}
+
+	scope, err := r.createScope(machineDeployment, ownedMachines)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create scope: %w", err)
+	}
+
+	if !r.hasUpgradeInstructions(scope) {
+		log.Info("MachineDeployment has no upgrade instructions, removing upgrade-to annotation from machines and stopping reconciliation")
```

Comment:
This would silently remove any manual upgrade from a single machine. right? Is this by design and we are fine with that?

Reply:
To be fair this was not mentioned in the proposal. But I did this intentionally to prevent unnecessary confusion and complexity later on (maybe I just added more confusion?). The rationale behind is that I think the users are supposed to use the `MachineDeployment` annotation and not the single machine ones. I can remove this if you think it's not the correct move.

Reply:
Users should be able to perform single machine operations, the orchestration should be optional. We would've directly implemented the upgrade logic in the higher level reconcilers such as `MachineDeployment` reconciler if the intention was to not perform single machine operations.